File: services/knowledge-service/src/main/java/com/poliwise/knowledge/entity/ProcessingJob.java
package com.poliwise.knowledge.entity;

import com.poliwise.knowledge.enums.ProcessingStatus;
import com.poliwise.knowledge.enums.ProcessingStep;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.OffsetDateTime;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "processing_jobs", schema = "knowledge")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class ProcessingJob {

    @Id
    @Column(name = "id", nullable = false)
    private UUID id;

    @Column(name = "document_id", nullable = false)
    private UUID documentId;

    @Enumerated(EnumType.STRING)
    @Column(name = "job_type", columnDefinition = "knowledge.processing_step")
    private ProcessingStep jobType;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", columnDefinition = "knowledge.processing_status")
    private ProcessingStatus status;

    @Column(name = "progress_percent")
    private Integer progressPercent;

    @Column(name = "started_at")
    private OffsetDateTime startedAt;

    @Column(name = "completed_at")
    private OffsetDateTime completedAt;

    @Column(name = "success")
    private Boolean success;

    @Column(name = "error_message", columnDefinition = "text")
    private String errorMessage;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "error_details", columnDefinition = "jsonb")
    private Map<String, Object> errorDetails;

    @Column(name = "retry_count")
    private Integer retryCount;

    @Column(name = "max_retries")
    private Integer maxRetries;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "output_metrics", columnDefinition = "jsonb")
    private Map<String, Object> outputMetrics;

    @Column(name = "created_at")
    private OffsetDateTime createdAt;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;


    /* ===== Helper methods cho pipeline ===== */

    public boolean isFinished() {
        return status == ProcessingStatus.READY || status == ProcessingStatus.FAILED;
    }

    public boolean canRetry() {
        if (retryCount == null || maxRetries == null) {
            return false;
        }
        return retryCount < maxRetries;
    }
}

File: services/ingestion-service/src/models/processing_job.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Text, BigInteger, Integer, Boolean
from sqlalchemy.dialects.postgresql import UUID as PG_UUID, JSONB
from src.db.session import Base


class ProcessingJob(Base):
    """SQLAlchemy model for knowledge.processing_jobs table."""
    __tablename__ = "processing_jobs"
    __table_args__ = {"schema": "knowledge"}

    id = Column(PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    document_id = Column(PG_UUID(as_uuid=True), nullable=False)
    document_version_id = Column(PG_UUID(as_uuid=True), nullable=True)
    job_type = Column(String(50), nullable=False)  # 'INGESTION' or 'REINDEX'
    status = Column(String(50), nullable=False)  # 'PENDING', 'PROCESSING', 'COMPLETED', 'FAILED'
    progress_percent = Column(Integer, default=0)
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    success = Column(Boolean, nullable=True)
    error_message = Column(Text, nullable=True)
    error_details = Column(JSONB, nullable=True)
    error_details = Column(JSONB, nullable=True)
    retry_count = Column(Integer, default=0)
    max_retries = Column(Integer, default=3)
    output_metrics = Column(JSONB, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    deleted_at = Column(DateTime, nullable=True)
